Issue #4377705: Create tar.gz from src.rpm

NirWolfer opened the pull request:
## Description
Released packed should be tar.gz instead of src.rpm, with a small change to the dir stucture
Pointers to the new tar.gz file are not being created by the script

##### What
Move libxlio.spec to the root of the directory being compressed.
Remove libxlio.spec from contrib/scripts/libxlio.spec

For now we will create both the src.rpm and the tar.gz
created in the build_pkg.sh , and moves it to the release folder.

##### Why ?
Issue #4377705

## Change type
What kind of change does this PR introduce?
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

## Check list
- [ ] Code follows the style de facto guidelines of this project
- [ ] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

NirWolfer commented:
build passing: https://nbuprod.blsm.nvidia.com/swx-media-master/job/libxlio/job/LibXLIO-release/55/
running rpmbuild on the created tar.gz passed (rpmbuild -ta /auto/sw/release/sw_acceleration/xlio/tmp/3.50.1/libxlio-3.50.1.tar.gz)

NirWolfer commented:
build passing: https://nbuprod.blsm.nvidia.com/swx-media-master/job/libxlio/job/LibXLIO-release/56/
running rpmbuild on the created tar.gz passed (rpmbuild -ta /auto/sw/release/sw_acceleration/xlio/tmp/3.50.3/libxlio-3.50.3.tar.gz)

NirWolfer commented:
bot:retest

dpressle commented on .ci/do_release.sh:
```diff
@@ -58,24 +84,29 @@ fi
-    DST_DIR=${release_folder}/${release_tag}
-
+    
+	if [ ! -d "${pkg_folder}" ]; then
```
Fix Indentation

NirWolfer replied:
Fixed

dpressle commented on .ci/do_release.sh:
```diff
@@ -58,24 +84,29 @@ fi
+    pushd "${release_folder}" || { echo "ERROR: Failed to pushd to ${release_folder}. Exit" ; exit 1; }
     sudo -E -u swx-jenkins ln -s "$DST_DIR/${pkg_name}" "${pkg_name}"
-    echo "Release found at $DST_DIR"
+	popd || { echo "ERROR: Failed to popd from ${release_folder}. Exit" ; exit 1; }
```
Fix Indentation

NirWolfer replied:
Fixed

dpressle commented on .ci/do_release.sh:
```diff
@@ -58,24 +84,29 @@ fi
+
+	# Creating tarball from src.rpm, and moving it to the release folder.
+	# src.rpm will be removed in the future.
+	make_tarball
```
why cant we run it right after we generate the src.rpm and finish all packages and then copy both is more stractured and less messy and hard to understand

MrBr-github replied:
+1

dpressle commented on .ci/do_release.sh:
```diff
@@ -18,6 +18,29 @@ print_help() {
+    echo "INFO: Making tarball for ${release_tag} with revision ${revision}"
+    tarball_name="libxlio-${release_tag}.tar.gz"
+	
+	mkdir -p /tmp/libxlio-"${release_tag}"
```
Fix indetation

NirWolfer replied:
Fixed

dpressle reviewed (changes requested):
Do we really need so much changes? cant it be simpler?

NirWolfer commented:
After discussing with @MrBr-github, we agreed that decoupling the tar.gz and src.rpm creation is better and will allow us to remove the src.rpm creation in the future when needed.

Changed to create both with the create package script, and copy both into the release directory

@dpressle regarding the revision in the name of the final tar.gz, i've tested it and since the folder inside does not contain the revision in it's name, the install command fails, i've reverted the name to include just the version

NirWolfer commented:
successful run: https://nbuprod.blsm.nvidia.com/swx-media-master/job/libxlio/job/LibXLIO-release/61/

dpressle commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +182,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
 fi
 
+# Create the final tarball separately after src.rpm generation
+if [ "$rc" -eq 0 ]; then
```
lets add option for generating tarball like we have for other options and maybe this code should only be running when $opt_srcpkg is selected

NirWolfer replied:
add a new option opt_tarpkg to create the tarball, invoked by --tarpkg|-t
add new flag to the if encapsulating this code

dpressle commented on .ci/do_release.sh:
```diff
@@ -43,39 +43,49 @@ if [ -z "${do_release}" ]; then
+    sudo -E -u swx-jenkins cp -v "${pkg_folder}/${tarball_name}" "${DST_DIR}"
 
-    cd "${release_folder}"
+    pushd "${release_folder}" || { echo "ERROR: Failed to pushd to ${release_folder}. Exit" ; exit 1; }
```
maybe do the cd before copy so if you cant cd dont copy and fail right away, otherwise you might have packages copied but no soft link

NirWolfer replied:
Moved the pushd to the start before any copy/semi link is made

NirWolfer commented:
bot:retest

NirWolfer commented:
bot:retest

MrBr-github commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
     rc=$((rc + $?))
 fi
 
+# Create the final tarball separately after src.rpm generation
```
Have you tested it without creation of the `src.rpm`?

NirWolfer replied:
No, but the creation of the original tar.gz which both the `src.rpm` and our `tar.gz` are created from is generated with `make dist` in line 127.

The creation of `src.rpm` is not related to it and shouldnt affect it

MrBr-github commented on .ci/do_release.sh:
```diff
@@ -43,39 +43,50 @@ if [ -z "${do_release}" ]; then
         exit 1
     fi
 
+	CWD=$(pwd)
```
CWD seems to be a bad variable name:
- After you changed directory CWD stops being _current working directory_ 
- Upper case used for GLOBAL or OS variables

NirWolfer replied:
changed CWD to files_dir

NirWolfer commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
 fi
 
+# Create the final tarball separately after src.rpm generation
+if [ $opt_tarpkg -eq 1 -a "$rc" -eq 0 ]; then
```
Following the same logic from the src.rpm creation

MrBr-github replied:
Seems to be problematic
How failure of building source rpm should affect tarball creation
see line 186

NirWolfer replied:
removed the check against `rc` in the if

NirWolfer commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
+    mkdir -p ${pkg_dir}/tarball_tmp >> ${pkg_log} 2>&1
+    tar -xzf ${pkg_tarball} -C ${pkg_dir}/tarball_tmp >> ${pkg_log} 2>&1
+    
+    # Move spec file to the root directory
```
those are the requirements for this change (see https://confluence.nvidia.com/pages/viewpage.action?spaceKey=SW&title=Packaging+tarballs+requirement)
other than specify that thats how they want their release tar struct to be im not sure what exactly to write here...

NirWolfer commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
 
+# Create the final tarball separately after src.rpm generation
+if [ $opt_tarpkg -eq 1 -a "$rc" -eq 0 ]; then
+    echo ${pkg_label} "Creating final tarball..."
```
changed the word `final` to `release`

NirWolfer commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
+    tarball_dir=${tarball_name%.tar.gz}
+    
+    # Create a temporary directory to work with the tarball contents
+    mkdir -p ${pkg_dir}/tarball_tmp >> ${pkg_log} 2>&1
```
Good point, changed from `mkdir -p ${pkg_dir}/tarball_tmp` to `tarball_tmp=$(mktemp -d -p ${pkg_dir})` and using that as the path

NirWolfer commented:
successful release build: https://nbuprod.blsm.nvidia.com/swx-media-master/job/libxlio/job/LibXLIO-release/64/

MrBr-github commented on contrib/build_pkg.sh:
```diff
@@ -182,6 +186,33 @@ if [ $opt_binpkg -eq 1 -a "$rc" -eq 0 ]; then
+if [ $opt_tarpkg -eq 1 -a "$rc" -eq 0 ]; then
+    echo ${pkg_label} "Creating final tarball..."
+    # Get version for directory name
+    tarball_name=$(basename ${pkg_tarball})
```
Please provide better name for `tarball_dir`
It's not clear why there are coexisting `pkg_dir` and  `tarball_dir`

NirWolfer replied:
name changed to `tarball_root_dir` to indicate the root directory inside the tarball.

galnoam commented:
/review

dpressle commented:
/improve

dpressle reviewed (changes requested):
Nir, please check pr-agent review and improve suggestions it has some interesting stuff

NirWolfer commented:
bot:retest
